src/librustc_typeck/check/method/probe.rs

```diff
@@ -1257,7 +1350,24 @@ impl<'a, 'gcx, 'tcx> ProbeContext<'a, 'gcx, 'tcx> {
     /// Find item with name `item_name` defined in impl/trait `def_id`
     /// and return it, or `None`, if no such item was defined there.
     fn associated_item(&self, def_id: DefId) -> Option<ty::AssociatedItem> {
-        self.fcx.associated_item(def_id, self.item_name)
+        match self.looking_for {
+            LookingFor::MethodName(item_name) => self.fcx.associated_item(def_id, item_name),
+            _ => None,
```

Comment:
Yeah, this method should go away. Callers can just use `impl_or_trait_item` instead. I've got a local branch doing that (might also try modifying `impl_or_trait_item` to return `impl Iterator`, since right now it only considers the first method that matches the return type...?)